src/group_theory/submonoid/inverses.lean

```diff
@@ -0,0 +1,163 @@
+/-
+Copyright (c) 2021 Andrew Yang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Andrew Yang
+-/
+
+import group_theory.submonoid.operations
+
+/-!
+
+# Submonoid of inverses
+
+Given a submonoid `N` of a monoid `M`, we define the submonoid `N.left_inv` as the submonoid of
+left inverses of `N`. When `M` is commutative, we may define `from_comm_left_inv : N.left_inv →* N`
+since the inverses are unique. When `N ≤ is_unit.submonoid M`, this is precisely
+the pointwise inverse of `N`, and we may define `left_inv_equiv : S.left_inv ≃* S`.
+
+For the pointwise inverse of submonoids of groups, please refer to
+`group_theory.submonoid.pointwise`.
+
+## TODO
+
+Define the submonoid of right inverses and two-sided inverses.
+
+-/
+section pointwise_inverse
+
+namespace submonoid
+
+@[to_additive] noncomputable
+instance {M : Type*} [monoid M] : group (is_unit.submonoid M) :=
+{ inv := λ x, ⟨_, (x.prop.unit⁻¹).is_unit⟩,
+  mul_left_inv := λ x, subtype.eq x.prop.unit.inv_val,
+ ..(show monoid (is_unit.submonoid M), by apply_instance) }
+
+
+@[to_additive] noncomputable
+instance {M : Type*} [comm_monoid M] : comm_group (is_unit.submonoid M) :=
+{ mul_comm := λ a b, mul_comm a b,
+ ..(show group (is_unit.submonoid M), by apply_instance) }
+
+section monoid
+
+variables {M : Type*} [monoid M] (S : submonoid M)
```

Comment:
No need to repeat `{M : Type*}`:
```suggestion
variables {M : Type*}

section pointwise_inverse

namespace submonoid

@[to_additive]
noncomputable instance [monoid M] : group (is_unit.submonoid M) :=
{ inv := λ x, ⟨_, (x.prop.unit⁻¹).is_unit⟩,
  mul_left_inv := λ x, subtype.eq x.prop.unit.inv_val,
 ..(show monoid (is_unit.submonoid M), by apply_instance) }


@[to_additive]
noncomputable instance [comm_monoid M] : comm_group (is_unit.submonoid M) :=
{ mul_comm := λ a b, mul_comm a b,
 ..(show group (is_unit.submonoid M), by apply_instance) }

section monoid

variables [monoid M] (S : submonoid M)
```

Reply:
You should also add the lemma:
```lean
@[to_additive] lemma is_unit.submonoid.coe_inv [monoid M] (x : is_unit.submonoid M) :
  ↑(x⁻¹) = (↑x.prop.unit⁻¹ : M) := rfl
```

Reply:
Should this be a simp lemma?

Reply:
I'm not sure, the RHS seems pretty messy